### ISSUE TITLE
Naive is not PAM

### ISSUE THREAD

**Issue by kno10** [ThrunGroup/BanditPAM#1]
https://github.com/ThrunGroup/BanditPAM/blob/3567dd2d49aadc4744710fb74069b1dda5a93730/src/kmedoids_ucb.cpp#L420-L434

The current code is O(N²k²) whereas the original PAM is only O(N²k) by computing the *change* in loss instead of recomputing the entire loss every time. The change can be computed efficiently when caching the distance to the nearest as well as second nearest medoid. Only for the 'winning' solution the second nearest needs to be updated at the end.

**Comment by motiwari** [comment]
Hi Prof. Schubert, thank you for filing this issue! We'll get this fixed.

@esfrankel can you take a look?

**Comment by motiwari** [comment]
@zaryabmakram could you take a look at this?

The core difference between PAM (which is O(kn^2)) and FastPAM (which is O(n^2)) is that the comparison to each potential medoid for the re-assignment is not necessary; rather, the best and second-best medoids and their distances are cached, which is used to do the comparison quickly. The original paper is at https://arxiv.org/pdf/1810.05691.pdf and the core logic in the Python implementation is [here](https://github.com/motiwari/BanditPAM-python/blob/master/data_utils.py#L258-L269).

**Comment by motiwari** [comment]
@mailology this is a good next issue for you, at a significantly more complex level. To resolve this, you'll need to understand both PAM and FastPAM1 from https://arxiv.org/pdf/1810.05691.pdf.

The core difference between PAM (which is O(kn^2)) and FastPAM (which is O(n^2)) is that the comparison to each potential medoid for the re-assignment is not necessary; rather, the best and second-best medoids and their distances are cached, which is used to do the comparison quickly. The original paper is at https://arxiv.org/pdf/1810.05691.pdf and the core logic in the Python implementation is here.

Let me know if you want to discuss the algorithms, since they are more complex conceptually than the issues we've been working on so far.

**Comment by kno10** [comment]
The original PAM algorithm *already* used this caching to reduce the cost to O(N²k). The current code in BanditPAM is O(N²k²) as you can easily tell from the four nested loops that have `k*N*N*k` iterations:
https://github.com/ThrunGroup/BanditPAM/blob/c5f9760e7c4a59b9381119545897adcd6516f06a/src/kmedoids_ucb.cpp#L347-L354
Reducing this to O(N²k) (the original PAM algorithm) is a first step, FastPAM in O(N²+Nk) then a second step.

FastPAM (please implement the newer version from the journal version!) computes k swaps in parallel at the same cost as computing one (with a case distinction in 2=O(1) cases, using O(k) additional memory), which makes it O(k) times faster.

The journal version of FastPAM is open access in the Information Systems journal:
> Erich Schubert, Peter J. Rousseeuw:
> Fast and eager k-medoids clustering: O(k) runtime improvement of the PAM, CLARA, and CLARANS algorithms.
> Inf. Syst. 101: 101804 (2021)
> https://doi.org/10.1016/j.is.2021.101804

Algorithm 2 is the classic PAM. Pay attention to Equation (5) to compute the loss change Delta, which uses the cached values of the nearest cluster, the distance to the nearest and the distance to the second nearest. This replaces the innermost loop! These need to be updated after each swap, not recomputed for each candidate. FastPAM then uses an array to compute k loss changes in O(N) steps.

FasterPAM (Faster = Fast and Eager) then reduces the number of iterations by eagerly performing any swap that improves the result. It still is O(N²) per iteration, but with fewer iterations.

To test your code, you can compare the results with respect to quality and run time to the Rust crate "kmedoids".

**Comment by motiwari** [comment]
Hi Prof. Schubert (@kno10), thank you for your comments and my apologies for the delay in getting back to you.

Your points are well taken -- and yes, you have indeed identified a bug in the linked code. By saving the assignments of each point, we should remove a factor of `k` (because we do not have to scan the medoids to compute the assignments) and the complexity of PAM becomes `O(N^2 * k)`. 

And yes, we will implement FastPAM with the additional optimization of keeping the second-best medoids for each point to reduce the complexity to `O(N^2 + Nk)`.

In fact, we implemented both of these optimizations in the [reference Python implementation](https://github.com/motiwari/BanditPAM-python/blob/master/data_utils.py#L275-L337) but missed them in the C++ implementation.

I will work with @mailology to implement these changes and we will also include instructions to cite the journal paper for people who use it.

**Comment by motiwari** [comment]
@kno10 thanks for the comments above; we made PAM `O(kn^2)` by replacing the innermost loop with the cached medoid distance in the PR above. 

We have opened a separate issue for the next step, the FastPAM implementation in #106. Let me know if you have any additional questions or comments.
